### ISSUE TITLE
Long-term features

### ISSUE THREAD

**Issue by SnirkImmington** [Immington-Industries/way-cooler#15]
These are the list of features we expect to support prior to a 1.0 release:&

- [x] i3 style tiling with custom layout options _(in progress)_ and workspace configuration (see #19)
- [x] Config files for Lua, fine-tuned global settings(see #18)
- [x] Inter-process channels to allow external programs to interact with way-cooler (see #20)
- [x] Lua thread running internally as a basic client (see #16)
- [x] An internal representation of the config JSON, accessed through Lua and IPC (see #17)

In addition, these external programs are needed:
- [ ] way-cooler-client: Simple client program to explore IPC and eval Lua in way-cooler from the command line
- [ ] way-cooler-bar: Simple and customizeable top bar powered by cairo with Lua widgets
- [ ] "barly": top bar with HTML/JS widgets
- [ ] way-cooler-helm: A launcher to be paired with way-cooler (optional)

Also, it would be good to have libraries for using way-cooler IPC from these languages:
- [ ] Lua
- [x] Python
- [ ] Rust

Finally, although it will be a difficult undertaking, we would like to provide way-cooler compatible versions of existing X programs/configuration:
- [ ] xbindkeys
- [ ] xdotool
- [ ] xmodmap
- [ ] xrandr

**Comment by SnirkImmington** [comment]
Updated to reflect Lua thread as an internal IPC client.

**Comment by agauniyal** [comment]
may I ask, why electron for bar? unless underlying engine would be shared by multiple widgets(or multiple bars/windows), you are looking at a v8 engine running just for bar?

**Comment by SnirkImmington** [comment]
I think we may have miscommunicated in our first Reddit post and possibly our readme. The top bar isn't a core part of the program, it's swappable. I had always said (but now I realize it's not documented anywhere) that we'd have a cairo bar (i.e. awesome's) with Lua widgets. But it's 2016 so Javascript is already far along in its crusade to be everywhere. 

The underlying engine _would_ be shared by all the widgets, and could probably extend to a bar and a dock. HTML and CSS are really great, some of the best UI and UX has been implemented with them. "Learn Javascript" tutorials are all over the web and I thought it would be nice for people familiar with them to customize their window manager with. Atom, for example, has seen a lot of success from this. However, this does come at the cost of running a full browser engine (as far as I can tell) to support.

TL; DR You don't need the electron bar. Javascript sucks; everyone uses it. We'll ship with a more lightweight bar (probably what more tiling wm users want) and you can use that one. The JS bar was just further along at the time of our post.

**Comment by Timidger** [comment]
Should we move this out of an issue and into a file in the root of the directory? Or would it be better to keep it in this format?

**Comment by SnirkImmington** [comment]
I like the idea of the big progress bar so we could move it to Trello or something
